hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ListIterator.java

```diff
@@ -0,0 +1,352 @@
+        }
+        if (StringUtils.isBlank(startKey)) {
+          // startKey is null or empty, then the seekKeyInDB="1024/"
+          if (cacheKey.startsWith(prefixKey)) {
+            cacheKeyMap.put(cacheKey, cacheOmInfo);
+          }
+        } else {
+          // startKey not empty, then the seekKeyInDB="1024/b" and
+          // seekKeyInDBWithOnlyParentID = "1024/". This is to avoid case of
+          // parentID with "102444" cache entries.
+          // Here, it has to list all the keys after "1024/b" and requires >=0
+          // string comparison.
+          if (cacheKey.startsWith(prefixKey) &&
+              cacheKey.compareTo(startKey) >= 0) {
+            cacheKeyMap.put(cacheKey, cacheOmInfo);
+          }
+        }
+      }
+    }
+
+    public boolean hasNext() {
+      return cacheCreatedKeyIter.hasNext();
+    }
+
+    public HeapEntry next() {
+      Map.Entry<String, Value> entry = cacheCreatedKeyIter.next();
+      return new HeapEntry(this.entryIteratorId, this.tableName,
+          entry.getKey(), entry.getValue());
+    }
+
+    public void close() {
+      // Nothing to close here
+    }
+  }
+
+  /**
+   * Implement lexicographical sorting of the file status by sorting file status
+   * across multiple lists. Each of these lists are sorted internally.
+   *
+   * This class implements sorted output by implementing a min heap based
+   * iterator where the initial element from each of sorted list is inserted.
+   *
+   * The least entry is removed and the next entry from the same list from
+   * which the entry is removed is added into the list.
+   *
+   * For example
+   * RawDir   - a1, a3, a5, a7
+   * RawFile  - a2, a4, a6, a8
+   *
+   * Min Heap is initially composed of {(a1, RawDir), (a2, RawFile)}
+   * THe least element is removed i.e a1 and then next entry from RawDir
+   * is inserted into minheap resulting in {(a2, RawFile), (a3, RawDir)}
+   *
+   * This process is repeated till both the lists are exhausted.
+   */
+  public static class MinHeapIterator implements ClosableIterator {
+    private final PriorityQueue<HeapEntry> minHeap = new PriorityQueue<>();
+    private final ArrayList<ClosableIterator> iterators = new ArrayList<>();
+
+    MinHeapIterator(OMMetadataManager omMetadataManager, String prefixKey,
+                    BucketLayout bucketLayout, String startKey,
+                    String volumeName, String bucketName) throws IOException {
+
+      this(omMetadataManager, prefixKey, startKey, volumeName,
+          bucketName, omMetadataManager.getDirectoryTable(),
+          omMetadataManager.getKeyTable(bucketLayout));
+    }
+
+    MinHeapIterator(OMMetadataManager omMetadataManager, String prefixKey,
+                    String startKey, String volumeName, String bucketName,
+                    Table... tables) throws IOException {
+      omMetadataManager.getLock().acquireReadLock(BUCKET_LOCK, volumeName,
+          bucketName);
+      try {
+        int iteratorId = 0;
```

Comment:
It would be better to use Something like `CACHE` or `DB` kind of enums which was previously used for better readability.

Reply:
We cannot use just one ENUM since there could be multiple table source.

Reply:
Just `CACHE` and `DB` don't need to be table specific. Generic ENUM should be fine

Reply:
We are using the same thing for file table and dir table for listing. We cannot simply do with 2 enums. The iterator id is indicator of the which iterator when has to read from to get the next entry. Heap will have only 1 entry from each iterator.